### PR TITLE
[download_client] Case-insensitive parsing of http headers

### PR DESCRIPTION
**Pull request by brendanyounger**
Use strcasestr() to case-insensitively find the Content-Range or Content-Length headers.  Also, fall back to the Content-Length header in case the server does not respond with Content-Range.

### CODE REVIEW

File: subsys/net/lib/download_client/src/download_client.c [hunk 1]
```diff
@@ -14,6 +14,11 @@
 #include <net/download_client.h>
 #include <logging/log.h>
 
+#if ! __GNU_VISIBLE
+/* strcasestr exists in libc and can be linked, but __GNU_VISIBLE must be 1 to be available in string.h */
+char	*strcasestr (const char *, const char *);
```
Comment (rlubos) [inline subsys/net/lib/download_client/src/download_client.c]: `strcasestr` is a non-standard function and is not available in Zephyr's minimal libc implementation (Zephyr default). In result, netither of our FOTA samples builds with this patch.

I think therefore that we should approach the problem in some other way. Normalize the header case before parsing it perhaps?

**Comment by lemrey** [comment]
Hi @brendanyounger, thank you for your patch.
We should indeed be able to handle lower-case header fields, since they are case insensitive, but like it's been already pointed out `strcasecmp` is not available in the C libraries that we support in NCS/Zephyr. A simpler approach like suggested could be to convert the header to lower case.

Regarding handling `Content-Length`, how would we determine the size of the whole file (to know when to stop)? At some point I had the idea of starting the whole download process with a `HEAD` request but dropped the idea because ultimately it made the logic of the thread more complicated.

**Comment by brendanyounger** [comment]
I believe I have addressed the style issues as well as using strcasestr.

File: subsys/net/lib/download_client/src/download_client.c [hunk 2]
```diff
@@ -274,6 +275,40 @@ static int get_request_send(struct download_client *client)
 	return 0;
 }
 
+static int my_strncasecmp(const char *s1, const char *s2, size_t n)
+{
+	if (n != 0) {
+		do {
+			if (tolower(*s1) != tolower(*s2++))
+				return (tolower(*s1) - tolower(*--s2));
```
Comment (rlubos) [inline subsys/net/lib/download_client/src/download_client.c]: Zephyr coding standard requires to add braces even for single-line code blocks. Please fix wherever applicable.

**Comment by rlubos** [comment]
@lemrey Please revisit at convenience.

**Comment by lemrey** [comment]
Have a few remarks:
- Since we check for other HTTP header fields as well, a better approach could be to lowercase the whole header and stick to `strstr`. We won't have to add those two functions, and we would handle correctly other fields too.

- It is possible to see if the server supports partial range requests from the HTTP status in the response, as described [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests). `200 OK` means the server did not support that feature. Perhaps we could check for the HTTP response status first and then `strstr` for the appropriate HTTP header field.

- Lastly, if the server doesn't support partial range requests they will only send one HTTP header followed by the entire payload. How is this going to work when after receiving a whole fragment we send another HTTP header to request the next bytes?

**Comment by carlescufi** [comment]
Note: This is an automated message.
@brendanyounger this Pull Request has not been updated for more than a month, please close it if not applicable anymore.
@lemrey  please close it if the author is not available

**Comment by rlubos** [comment]
Stale

**Pull request closed without merging**